File: packages/interviewbit/programming/binary-search/matrix-search/matrix-search.ts
/**
 * MatrixSearch
 */

function matrixSearch(matrix: number[][], target: number): number {
  const m: number = matrix.length;
  if (m === 0) {
    return 0;
  }
  const n: number = matrix[0].length;
  let start: number = 0;
  let end: number = m * n - 1;

  while (start <= end) {
    // tslint:disable:no-bitwise
    const middle: number = (start + end) >>> 1;
    const row: number = Math.floor(middle / n);
    const col: number = middle % n;

    if (matrix[row][col] < target) {
      start = middle + 1;
    } else if (matrix[row][col] > target) {
      end = middle - 1;
    } else {
      return 1;
    }
  }

  return 0;
}

export { matrixSearch };

File: packages/interviewbit/programming/binary-search/matrix-search/matrix_search.py
# pylint: disable=missing-docstring

def matrixSearch(matrix, target):
  m = len(matrix)
  if m == 0:
    return 0

  n = len(matrix[0])
  start = 0
  end = m * n - 1

  while start <= end:
    middle = (start + end) // 2
    row = middle // n
    col = middle % n
    if matrix[row][col] < target:
      start = middle + 1
    elif matrix[row][col] > target:
      end = middle - 1
    else:
      return 1

  return 0
